docs/metrics.asciidoc

```diff
@@ -60,3 +60,69 @@ This value is normalized by the number of CPU cores and it ranges from 0 to 100%
 
 The Resident Set Size,
 the amount of memory the process occupies in main memory (RAM).
+
+[float]
+[[metric-nodejs.active_handles]]
+=== `nodejs.active_handles`
+
+* *Type:* Long
+* *Format:* Counter
+
+The number of active libuv handles,
+likely held open by currently running I/O operations.
+
+[float]
+[[metric-nodejs.active_requests]]
+=== `nodejs.active_requests`
+
+* *Type:* Long
+* *Format:* Counter
+
+The number of active libuv requests,
+likely waiting for a response to an I/O operation.
+
+[float]
+[[metric-system.process.cpu.user.ticks]]
+=== `system.process.cpu.user.ticks`
```

Comment:
Do you typically use user/system ticks metrics when diagnosing Node.js application issues?

I'm wondering if we should perhaps leave them out for now, and if it's useful then later add it across all agents. The `system.process.cpu.total.norm.pct` is already a combination of user and system time. Reporting "ticks" doesn't seem very friendly, I think we should convert it to an amount of time, or perhaps a percentage.

Reply:
The split is valuable to surface the difference between slowness due to application execution time and slowness due to async requests overloading the system cpu time.

As for using something friendlier than a "ticks" metric, I just used ticks because that is the only format that data is available in Node.js. We'd have to do some extra work to identify the CPU capabilities and do some extra math to report as time or percentage. Not impossible, but non-trivial.

Reply:
Also, when we looked, "ticks" was the only metric used for user and system CPU by metricbeat, and we didn't feel comfortable at the time to invent a new metric.

Reply:
> As for using something friendlier than a "ticks" metric, I just used ticks because that is the only format that data is available in Node.js. We'd have to do some extra work to identify the CPU capabilities and do some extra math to report as time or percentage. Not impossible, but non-trivial.

Reporting as a time might be a little tricky in Node.js if it means having to use `sysconf`, but I think it would be possible to avoid this by building in assumptions about supported architectures. That's what we do in Go, but maybe there's some architecture supported by Node.js that's not supported by Go that would pose a problem. I'm not sure, and on second thought reporting as time probably isn't easier to grok anyway.

Reporting as a percentage should be pretty straightforward. We're already doing this for the total CPU time (`system.process.cpu.total.norm.pct`), which is a combination of user+system. Decomposing that metric would give you `...user.norm.pct` and `...system.norm.pct`.

> Also, when we looked, "ticks" was the only metric used for user and system CPU by metricbeat, and we didn't feel comfortable at the time to invent a new metric.

 Fair point. Metricbeat currently provides `system.cpu.user.norm.pct`, but not `system.process.user.norm.pct`. I think it would be easier to understand, hence reasonable to add it, but we can defer that discussion to ecs-metrics.

Reply:
I would prefer a percentage as well. Do you think it's worth spending time on it in the PR? Or can we just add it later while keeping the ticks metric? Or would it be better to add it now in case we then want to remove the ticks metrics which would be a breaking change?

Reply:
To calculate the percentage, I'd need the total tick count too. @watson do you know of a way to get that?

Reply:
Ah, because total isn't just `user` + `system`? I don't know that unfortunately... will have to look more into it to be sure.

Reply:
@watson I think we go for eventual consistency: report percentages now, and then coordinate with Metricbeat to add those metrics later. I wouldn't add in ticks if we have percentages.

@Qard the `utime` and `stime` values from `/proc/pid/stat` are in ticks, which is what we're using to calculate the total process CPU usage in https://github.com/elastic/apm-agent-nodejs/blob/master/lib/metrics/platforms/linux/stats.js#L160 (`user`+`system`). I think we should just move the user/system breakdown to that code, and not use `process.cpuUsage`.

Reply:
Made a commit switching to percent system and cpu time, using the proc file stuff @axw mentioned.